src/core/tsi/ssl_transport_security_utils.h

```diff
@@ -0,0 +1,147 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_TSI_SSL_TRANSPORT_SECURITY_UTILS_H
+#define GRPC_CORE_TSI_SSL_TRANSPORT_SECURITY_UTILS_H
+
+#include <grpc/support/port_platform.h>
+
+#include <openssl/x509.h>
+
+#include "absl/strings/string_view.h"
+
+#include <grpc/grpc_security_constants.h>
+
+#include "src/core/tsi/ssl/key_logging/ssl_key_logging.h"
+#include "src/core/tsi/transport_security_interface.h"
+
+namespace grpc_core {
+
+// Converts an SSL error status code to a readable string.
+//
+// error: the SSL error status code.
+//
+// return: the corresponding status string.
+const char* SslErrorString(int error);
+
+// Logs the SSL error stack.
+void LogSslErrorStack(void);
+
+// Performs an SSL_write and handle errors.
+//
+// ssl: the SSL object to write to.
+// unprotected_bytes: the buffer containing the bytes for writing to |ssl|.
+// unprotected_bytes_size: the size of the buffer |unprotected_bytes|.
+//
+// return: TSI_OK if the write operation succeeds or corresponding TSI errors.
+tsi_result DoSslWrite(SSL* ssl, unsigned char* unprotected_bytes,
+                      size_t unprotected_bytes_size);
+
+// Performs an SSL_read and handle errors.
+//
+// ssl: the SSL object to read from.
+// unprotected_bytes: the buffer to which this function will populate the read
+//                    result from |ssl|.
+// unprotected_bytes_size: the maximum size of the buffer |unprotected_bytes|.
+//                         This will be populated with the size of the bytes
+//                         read from |ssl| if this function returns TSI_OK.
+//
+// return: TSI_OK if the write operation succeeds or corresponding TSI errors.
+tsi_result DoSslRead(SSL* ssl, unsigned char* unprotected_bytes,
```

Comment:
Would it make sense to change this to use some idiomatic C++ type for the buffers?  It would be nice, for example, to have some type that encodes both the buffer space and the size in a single object, similarly to how `std::string` works.  But I don't know if it makes sense to do that in this PR, given that the surrounding code is presumably still using C-style buffers.

Reply:
It would be great if the function can take a modern STL container (e.g., a `std::vector<uint8_t>`). However, this change (I think) requires changes all the way to the internal structure of TSI, which is out of scope of this PR, which aims to refactor ssl frame protector only.